### PR TITLE
Kernel/aarch64: Implement TimeManagement

### PR DESCRIPTION
**Pull request by FireFox317**
By implementing the `TimeManagment` class for aarch64, we now keep track of the elapsed time, which allows us to prepend the correct time since boot for the format (`dmesgln` etc) messages, and actually call the `Scheduler::timer_tick` every 4ms!

image

### CODE REVIEW

File: Kernel/Arch/aarch64/Processor.h [hunk 1]
```diff
@@ -186,6 +186,9 @@ class Processor {
 
 private:
     u32 m_in_critical { 0 };
+
+    // FIXME: This shouldn't be a FlatPtr
+    FlatPtr m_in_irq { 0 };
```
Comment (nico) [inline Kernel/Arch/aarch64/Processor.h:191]: Does anything ever write to this?

**Review by nico — APPROVED** [review]
Looks good to me after comments, thanks!

File: Kernel/Arch/aarch64/Processor.h (hunk 1 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Processor.h:192]: Nope not yet, the current code has no way to differentiate between irqs, fiqs and synchronous exceptions (syscalls). I've added that comment to the FIXME.

File: Kernel/Arch/aarch64/RPi/Timer.cpp [hunk 2]
```diff
@@ -52,23 +59,48 @@ u64 Timer::microseconds_since_boot()
...
-    m_current_timer_value += m_interrupt_interval;
-    set_compare(TimerID::Timer1, m_current_timer_value);
+    // Should only be called by the time keeper interrupt handler!
+    u64 current_value = microseconds_since_boot();
+    u64 delta_ticks = m_main_counter_drift;
+    if (current_value >= m_main_counter_last_read) {
+        delta_ticks += current_value - m_main_counter_last_read;
+    } else {
+        // the counter wrapped around
+        delta_ticks += m_main_counter_last_read - current_value;
```
[the remark this answers is not in the document]
Comment (FireFox317) [reply Kernel/Arch/aarch64/RPi/Timer.cpp]: Yep, you are correct, that is totally bogus indeed. I guess we never ran Serenity long enough to overflow the 64-bit counter :laughing:
I copied this from the HPET code, which has the same bug. I've updated both places (should really share this code, but that can be improved later)
https://github.com/SerenityOS/serenity/blob/8b5b42f667527fe9d304cf597907897cce4a3d32/Kernel/Arch/x86/Time/HPET.cpp#L256-L285

**Review by gunnarbeutner — APPROVED** [review]

**Pull request merged**